core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef __wasi__
+#error This example only compiles to WASM/WASI target
+#endif
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <math.h>
+
+#if USE_CUSTOM_SYNC_PRIMITIVES != 0
+#include "sync_primitives.h"
+#else
+#include <pthread.h>
+#endif
+
+#include "wasi_thread_start.h"
+
+enum CONSTANTS {
+    NUM_ITER = 1000000,
+    NUM_RETRY = 5,
+    SECOND = 1000 * 1000 * 1000, /* 1 second */
+    TIMEOUT = 50LL * SECOND,
+    MAX_NUM_THREADS = 4,
+};
+
+typedef struct {
+    start_args_t base;
+    int th_done;
+    int thread_id;
+    int value;
+} shared_t;
+
+unsigned int factorised_number = 1;
+unsigned int prime_numbers = 0;
+double percentage = 0.1;
+
+// Inited threads - Deinited threads. Should always be zero in the end. Otherwise it's a leak.
+int thread_init_delta = 0;
+
+pthread_mutex_t mutex;
+shared_t threads[MAX_NUM_THREADS];
+
+bool is_prime(unsigned int num) {
+    for (unsigned int  i = 2; i <= (unsigned int)(sqrt(num)); ++i) {
+        if (num % i == 0) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void
+__wasi_thread_start_C(int thread_id, int *start_arg)
+{
+    shared_t *data = (shared_t *)start_arg;
+
+    if (is_prime(data->value)) {
+        __atomic_fetch_add(&prime_numbers, 1, __ATOMIC_SEQ_CST);
+    }
+
+    __atomic_store_n(&data->th_done, 1, __ATOMIC_SEQ_CST);
+    __builtin_wasm_memory_atomic_notify(&data->th_done, 1);
+}
+
+unsigned int validate() {
+    unsigned int counter = 0;
+    for (unsigned int i = 2; i <= factorised_number; ++i) {
+        counter += is_prime(i);
+    }
+
+    return counter;
+}
+
+int
+main(int argc, char **argv)
+{
+    pthread_mutex_init(&mutex, NULL);
+
+    for (int i = 0; i < MAX_NUM_THREADS; ++i) {
```

Comment:
Should this loop be moved inside the while loop to avoid dupplication?

Reply:
I'd say they're a bit different anyway. 

e.g. here we don't need to cleanup the resources, wait for the thread, etc. 
I think moving it inside the next loop might introduce even more complexity